contracts/citycoin-core.clar

```diff
@@ -1,201 +1,539 @@
+(define-constant TOKEN_HALVING_BLOCKS u210000)
+
+;; store block height at each halving, set by register-user    
+(define-data-var coinbaseThreshold1 uint u0)
+(define-data-var coinbaseThreshold2 uint u0)
+(define-data-var coinbaseThreshold3 uint u0)
+(define-data-var coinbaseThreshold4 uint u0)
+(define-data-var coinbaseThreshold5 uint u0)
+
+;; return coinbase thresholds if contract activated
+(define-read-only (get-coinbase-thresholds)
+  (let
+    (
+      (activated (var-get activationReached))
+    )
+    (asserts! activated (err ERR_CONTRACT_NOT_ACTIVATED))
+    (ok {
+      coinbaseThreshold1: (var-get coinbaseThreshold1),
+      coinbaseThreshold2: (var-get coinbaseThreshold2),
+      coinbaseThreshold3: (var-get coinbaseThreshold3),
+      coinbaseThreshold4: (var-get coinbaseThreshold4),
+      coinbaseThreshold5: (var-get coinbaseThreshold5)
+    })
+  )
 )
 
-;; helper function
-(define-read-only (is-between (value uint) (low uint) (high uint))
-  (and 
-    (>= value low)
-    (<= value high)
+;; function for deciding how many tokens to mint, depending on when they were mined
+(define-read-only (get-coinbase-amount (minerBlockHeight uint))
+  (begin
+    ;; if contract is not active, return 0
+    (asserts! (>= minerBlockHeight (var-get activationBlock)) u0)
+    ;; if contract is active, return based on issuance schedule
+    ;; halvings occur every 210,000 blocks for 1,050,000 Stacks blocks
+    ;; then mining continues indefinitely with 3,125 CityCoins as the reward
+    (asserts! (> minerBlockHeight (var-get coinbaseThreshold1))
+      (if (<= (- minerBlockHeight (var-get activationBlock)) u10000)
+        ;; bonus reward first 10,000 blocks
+        u250000
+        ;; standard reward remaining 200,000 blocks until 1st halving
+        u100000
+      )
+    )
+    ;; computations based on each halving threshold
+    (asserts! (> minerBlockHeight (var-get coinbaseThreshold2)) u50000)
+    (asserts! (> minerBlockHeight (var-get coinbaseThreshold3)) u25000)
+    (asserts! (> minerBlockHeight (var-get coinbaseThreshold4)) u12500)
+    (asserts! (> minerBlockHeight (var-get coinbaseThreshold5)) u6250)
+    ;; default value after 5th halving
+    u3125
   )
 )
 
-;; Initialize
-(map-set ActiveContracts "logic" .citycoin-logic-v1)
+;; mint new tokens for claimant who won at given Stacks block height
+;; TODO: verify access to function only available by core
+(define-private (mint-coinbase (recipient principal) (stacksHeight uint))
+  (contract-call? .citycoin-token mint (get-coinbase-amount stacksHeight) recipient)
+)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; UTILITIES
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; check if contract caller is city wallet
+(define-private (is-authorized-city)
+  (is-eq contract-caller (var-get cityWallet))
+)
+
+;; check if contract caller is contract owner
+(define-private (is-authorized-owner)
+  (is-eq contract-caller CONTRACT_OWNER)
```

Comment:
I would say can stay static. The only thing that this function secures is full contract initialization, that will be executed right after deployment.

Reply:
Yes so far this is only used for the initial deployment to avoid a circular dependency between the core and logic contracts. The city wallet and the upcoming auth contract in #119 will be the main methods for making any changes within the contract.